### PR TITLE
reject when request method is not a listed method

### PR DESCRIPTION
**Pull request by mattking17**
This covers step 20 on Request initialization:

> If init's method member is present, let method be it and run these substeps:
>    If method is not a method or method is a forbidden method, throw a TypeError.

### CODE REVIEW

**Comment by domenic** [comment]
fetch() shouldn't throw; it should reject. new Request should throw though.

(This is implicit in the spec. Since fetch() returns a Promise WebIDL type, it automatically catches thrown errors in the algorithm and turns them into rejections.)

**Comment by mattking17** [comment]
I updated the patch, I think this is more in line with what you mentioned.

**Comment by domenic** [comment]
Ideally `new Request({ method: 'BLADH' })` should throw a TypeError... The code in this polyfill seems structured rather different from the spec though.

**Comment by mattking17** [comment]
Right...that's how it was originally in the patch, but maybe I misunderstood. If we throw on `new Request({ method: 'BLADH' })` then `window.fetch` polyfill will always throw when the `method` is bad, and will never get to the point of being able to reject.

I think it should just throw an error regardless of how it's called, instead of waiting for the request to be made.

**Comment by domenic** [comment]
No, there's a differnece between `new Request` throwing, and `fetch` throwing. Check browsers.

**Comment by mattking17** [comment]
This should cover throwing only `new Request...` but not when calling `window.fetch`

**Comment by domenic** [comment]
Looks good :). It's a shame the structure here is so different from the spec (e.g. there being a Request.prototype.fetch method). That's what makes this complicated.

File: fetch.js [hunk 1]
```diff
@@ -225,9 +225,14 @@
     this.mode = options.mode || null
     this.referrer = null
 
+    if (!options.constructedWithFetch && methods.indexOf(this.method) === -1) {
```
Comment (mislav) [inline fetch.js]: This consults `methods` array which is not a definite list of all supported HTTP methods, but only of few main methods that the spec defines as case-insensitive.

To properly implement this per-spec, we would have to maintain the [list of forbidden methods instead](https://fetch.spec.whatwg.org/#forbidden-method):

> A forbidden method is a method that is a byte-case-insensitive match for one of `CONNECT`, `TRACE`, and `TRACK`.

Furthermore, a ["method" is defined](https://tools.ietf.org/html/rfc7230#section-3.2.6) simply as a "token", which to me seems like it's basically any non-empty string.

So, we would have to check if the method is nonempty and if it's not listed among forbidden methods.

**Comment by dgraham** [comment]
Closing as stale for now. Feel free to reopen if addressing @mislav's [review comments](https://github.com/github/fetch/pull/132#discussion_r49263515).

**Pull request closed without merging**